packages/trimerge-sync/src/lib/LeaderManager.ts

```diff
@@ -0,0 +1,202 @@
+      action: 'request',
+      clientId: this.clientId,
+    });
+    this.electionTimeout = setTimeout(
+      () => this.finishElection(),
+      this.settings.electionTimeoutMs,
+    );
+  }
+
+  private cancelElection() {
+    if (this.electionTimeout) {
+      clearTimeout(this.electionTimeout);
+      this.electionTimeout = undefined;
+    }
+  }
+  private finishElection() {
+    this.electionTimeout = undefined;
+    const [proposedLeader] = Array.from(this.potentialLeaders).sort();
+    this.setLeader(proposedLeader);
+  }
+
+  private setLeader(leaderId: string | undefined) {
+    this.cancelElection();
+    const { clientId } = this;
+    this.currentLeaderId = leaderId;
+    const isLeader = leaderId === clientId;
+    if (this.isLeader !== isLeader) {
+      this.isLeader = isLeader;
+      this.onLeaderChange(isLeader);
+      if (this.leaderHeartbeat) {
+        clearInterval(this.leaderHeartbeat);
+        this.leaderHeartbeat = undefined;
+      }
+      if (isLeader) {
+        this.broadcastEvent({ type: 'leader', action: 'current', clientId });
+        this.leaderHeartbeat = setInterval(
+          () => this.onLeaderHeartbeat(),
+          this.settings.heartbeatMs,
+        );
+      }
+    }
+    if (this.heartbeatTimeout) {
+      clearTimeout(this.heartbeatTimeout);
+      this.heartbeatTimeout = undefined;
+    }
+    if (!isLeader) {
+      this.heartbeatTimeout = setTimeout(
+        () => this.onHeartbeatTimeout(),
+        this.settings.heartbeatTimeoutMs,
+      );
+    }
+  }
+
+  private onLeaderHeartbeat() {
+    const { clientId } = this;
+    this.broadcastEvent({ type: 'leader', action: 'current', clientId });
+  }
+  private onHeartbeatTimeout() {
+    this.heartbeatTimeout = undefined;
+    this.elect();
+  }
+
+  receiveEvent({ action, clientId: otherClientId }: LeaderEvent) {
+    const { currentLeaderId, clientId, closed } = this;
+    if (closed) {
+      return;
+    }
+    switch (action) {
+      case 'request':
+        if (currentLeaderId === undefined) {
+          // An election is needed
+          this.elect();
+          this.potentialLeaders.add(otherClientId);
+          this.broadcastEvent({
+            type: 'leader',
+            action: 'accept',
+            clientId,
+          });
+        } else if (currentLeaderId === clientId) {
+          // We already are the leader
+          this.broadcastEvent({
+            type: 'leader',
+            action: 'current',
+            clientId,
+          });
+        }
+        break;
+      case 'current':
+        if (this.isLeader) {
+          // This will happen if there's a disconnect/messages are delayed
+          if (otherClientId < clientId) {
+            this.setLeader(undefined);
+          }
+          this.elect();
+        } else {
+          this.setLeader(otherClientId);
+        }
+        break;
+      case 'accept':
+        this.potentialLeaders.add(otherClientId);
+        break;
+      case 'withdraw':
```

Comment:
What happens if a client drops out (i.e., loses connection) but doesn't actually send a withdraw message? Will their id stick around in `potentialLeaders` forever? Is that a problem?

Reply:
potentialLeaders is cleared every time an election starts, so that should be OK, this delete is just for the scenario where a client drops out (cleanly) during an election